cpp/include/raft/random/rng.cuh

```diff
@@ -379,6 +383,82 @@ void sampleWithoutReplacement(const raft::handle_t& handle,
     rng_state, out, outIdx, in, wts, sampledLen, len, handle.get_stream());
 }
 
+/**
+ * @brief Sample the input array without replacement, optionally based on the
+ * input weight vector for each element in the array
+ *
+ * Implementation here is based on the `one-pass sampling` algo described here:
+ * https://www.ethz.ch/content/dam/ethz/special-interest/baug/ivt/ivt-dam/vpl/reports/1101-1200/ab1141.pdf
+ *
+ * @note In the sampled array the elements which are picked will always appear
+ * in the increasing order of their weights as computed using the exponential
+ * distribution. So, if you're particular about the order (for eg. array
+ * permutations), then this might not be the right choice!
+ *
+ * @tparam DataT data type (do not specify this explicitly,
+ *         as the compiler will deduce it from @c out )
+ * @tparam IdxT index type (do not specify this explicitly,
+ *         as the compiler will deduce it from the arguments)
+ * @tparam WeightsT weights type (defaults to @c double ,
+ *         so that passing in std::nullopt works)
+ *
+ * @param handle
+ * @param rng_state random number generator state
+ * @param out Array (of length 'sampledLen')
+ *        of samples from the input array
+ * @param outIdx If provided, array (of length 'sampledLen')
+ *        of the indices sampled from the input array
+ * @param in Input array to be sampled (of length 'len')
+ * @param wts Weights array (of length 'len').
+ *        If not provided, uniform sampling will be used.
+ */
+template <typename DataT, typename IdxT, typename WeightsT = double>
+void sampleWithoutReplacement(const raft::handle_t& handle,
+                              RngState& rng_state,
+                              raft::device_vector_view<DataT, IdxT> out,
```

Comment:
@cjnolet Does the `handle` still go first?  Also, would `rng_state` be considered an output (or inout) here?

Reply:
Yeah, `handle, `rng_state`, then `in`, `out`, `params`.

Reply:
The more I think about it- I guess it seemed intuitive to me intially to put `rng_state` right after handle but I guess it doesn't have to be that way. As long as we are consistent. The knn/ann APIs are in an open PR and I see the `rng_state` object as being similar to the `ann_index` object, so I guess we also put them after the ins/outs. Do you have a strong opinion on this?

Reply:
@cjnolet I don't have a strong opinion -- I don't mind it going right after `handle`, especially since most of the functions take `RngState&`.  It would be harder to find if we were to list it with the output parameters.

I've reordered the parameters; thanks!